### PR TITLE
feat: add pipe() for ForkAndExec

### LINKED ISSUE #22: pipe() syscall isn't covered by extrasafe
Hello again, I've been continuing work on my [sshield](https://github.com/gotlougit/sshield) program and have run into more issues with using `extrasafe`. More specifically, now `pipe()` or `pipe2()` syscalls aren't covered by `extrasafe`.

I think the ideal place for these to exist would be `builtins::danger_zone`, though I am not sure, thus creating this issue. I'd be happy to create a PR to get this functionality in the code.

### PR DESCRIPTION
**Pull request by gotlougit**
Fixes #22

### CODE REVIEW

File: src/builtins/danger_zone.rs [hunk 1]
```diff
@@ -97,6 +97,7 @@ impl RuleSet for ForkAndExec {
              Sysno::execve, Sysno::execveat,
              Sysno::wait4, Sysno::waitid,
              Sysno::clone, Sysno::clone3,
+             Sysno::pipe, Sysno::pipe2
```
Comment (boustrophedon) [inline src/builtins/danger_zone.rs:100]: Ah sorry I was unclear in my comment on the original issue.

You should make a separate method in this ruleset (or make a separate ruleset that's just called Pipes or something) which enables pipes separately from the rest. 

I think maybe just making a
```
pub struct Pipes; 
impl RuleSet for Pipes {
  [...] simple_rules [...] 
    vec![Sysno::pipe, Sysno::pipe2]
```

is probably the simplest.

Comment (gotlougit) [reply src/builtins/danger_zone.rs:100]: Thanks for the clarification. I'll adjust my PR accordingly

File: src/builtins/danger_zone.rs [hunk 2]
```diff
@@ -125,3 +125,20 @@ impl RuleSet for ForkAndExec {
         "ForkAndExec"
     }
 }
+
```
Comment (boustrophedon) [inline src/builtins/danger_zone.rs:128]: I/we should probably change the doc comment in ForkAndExec as well as here to first describe what this actually allows, then put the rest into a `# Security considerations` section.

I also don't think that the description is entirely accurate - I believe you can send file descriptors across unix domain sockets so presumably you can send the pipe's fd along those as well, and the other process can be anything.

**Comment by boustrophedon** [comment]
Thanks for the PR!

File: src/builtins/danger_zone.rs [hunk 3]
```diff
@@ -125,3 +125,23 @@ impl RuleSet for ForkAndExec {
...
+/// [`Pipes`] is in the danger zone because it can be used create a pipe for IPC.
+///
+/// # Security Considerations
+/// The piped process will still be under seccomp's restrictions (see
```
Comment (boustrophedon) [inline src/builtins/danger_zone.rs:132]: Sorry again for not getting to this faster. This still isn't correct even without sending fds via unix domain sockets.

You can, for example, create the pipe, fork the process, and then the parent and the child both have the pipe but even if you apply seccomp to the parent, it doesn't apply to the already-forked child (unless you use `apply_to_all_threads` of course)

Comment (gotlougit) [reply src/builtins/danger_zone.rs:132]: Sorry about that, I'll adjust the docs accordingly.

**Comment by boustrophedon** [comment]
I don't really want to go back and forth on this much more, can you please make the comment
```
/// [`Pipes`] allows you to create anonymous pipes via the `pipe` syscalls for inter-process
/// communication.
///
/// # Security considerations
/// The program on the other end of the pipe may not be under the same seccomp restrictions as the
/// current process.
```

and could you also then squash all the commits down? Thanks!

**Comment by boustrophedon** [comment]
Actually I'm wrong here - there really aren't any security considerations for allowing the pipe syscall because you need to call pipe prior to forking so that the forked process can have the copy, but that means that either you're doing it before calling seccomp, in which case there aren't any security considerations for extrasafe, or you're doing it after calling seccomp, in which case any forked processes will also be under the same seccomp restrictions.

So I think we can just remove the security considerations section entirely and maybe move it to its own module or something.

**Comment by gotlougit** [comment]
Cool, and sorry about the misphrasing of the comment

**Comment by boustrophedon** [comment]
Thanks!

**Pull request merged**